### PR TITLE
Make TashScheduler support requests priority.

### PR DESCRIPTION
**Pull request by kzvieriev**
Overloading EnqueueTask() method with priority input. Now enqueued
requests will be executed in priority order. Request with same
priority follow FIFO approach. thread::Priority is a helper enum,
NORMAL=500 is a default priority for tasks.

Resolves: OLPEDGE-2306

Signed-off-by: Kostiantyn Zvieriev <ext-kostiantyn.zvieriev@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/Common.h [hunk 1]
```diff
@@ -71,26 +71,50 @@ inline client::CancellationToken ScheduleFetch(TaskScheduler&& schedule_task,
...
+ * @param args Additional agrs to pass to TaskContext.
+ * @return CancellationToken used to cancel the operation.
+ */
+template <typename Function, typename Callback, typename... Args>
+inline client::CancellationToken AddTask(
+    const std::shared_ptr<thread::TaskScheduler>& task_scheduler,
+    const std::shared_ptr<client::PendingRequests>& pending_requests,
+    Function task, Callback callback, Args&&... args) {
+  return AddTask(task_scheduler, pending_requests, task, callback,
+                 thread::NORMAL);
```
Comment (mykhailo-kuchma) [inline olp-cpp-sdk-dataservice-read/src/Common.h:115]: args are missing

**Review by mykhailo-kuchma — CHANGES_REQUESTED** [review]

File: olp-cpp-sdk-core/tests/thread/ThreadPoolTaskSchedulerTest.cpp [hunk 2]
```diff
@@ -126,3 +129,129 @@ TEST(ThreadPoolTaskSchedulerTest, MultiUserPush) {
...
+  const auto start = chrono::system_clock::now();
+  auto check_condition = [&]() {
+    return counter < expected_tasks &&
+           chrono::duration_cast<chrono::seconds>(chrono::system_clock::now() -
+                                                  start) < chrono::seconds(10);
+  };
+
+  while (check_condition()) {
+    std::this_thread::sleep_for(kSleep);
+  }
```
Comment (andescu) [inline olp-cpp-sdk-core/tests/thread/ThreadPoolTaskSchedulerTest.cpp:184]: This is not a very good system as it created a lot of flakiness. Instead, why not push a last, very low priority task that sets a promise<void> and you just wait for the associated future?

File: olp-cpp-sdk-core/src/thread/ThreadPoolTaskScheduler.cpp [hunk 3]
```diff
@@ -59,9 +62,41 @@ void SetCurrentThreadName(const std::string& thread_name) {
...
+struct PrioritizedTask {
+  TaskScheduler::CallFuncType function;
+  uint32_t priority;
+};
+
+struct ComparePrioritizedTask {
+  bool operator()(const PrioritizedTask& lhs,
+                  const PrioritizedTask& rhs) const {
+    return lhs.priority < rhs.priority;
+  }  // namespace
```
Comment (andescu) [inline olp-cpp-sdk-core/src/thread/ThreadPoolTaskScheduler.cpp:74]:
```suggestion
  }
```

Comment (kzvieriev) [reply olp-cpp-sdk-core/src/thread/ThreadPoolTaskScheduler.cpp]: Oops, not even sure how that happened, sorry. Fixed! :-)

**Review by andescu — APPROVED** [review]

**Review by mykhailo-kuchma — APPROVED** [review]

**Pull request merged**